File: sfml demos/exercise 04/SFML/UIButton.cpp
#include "UIButton.h"


UIButton::UIButton(string normalText, string clickedText, Vector2f position)
{
	try
	{
		if (!normalTexture.loadFromFile(normalText))
			cout << "Error loading file" << endl;
		
		if (!clickedTexture.loadFromFile(clickedText))
			cout << "Error loading file" << endl;

		this->normal.setTexture(normalTexture);
		this->clicked.setTexture(clickedTexture);
		this->normal.setPosition(position);
		this->clicked.setPosition(position);
		setState(false);
	}
	catch (int e)
	{
		cout << "ERROR" << endl;
	}
}

UIButton::~UIButton()
{
}

void UIButton::checkClick(Vector2f mousePos)
{
	int width = currentSprite.getGlobalBounds().width;
	int height = currentSprite.getGlobalBounds().height;

	if (mousePos.x > currentSprite.getPosition().x && mousePos.x < (currentSprite.getPosition().x + width) &&
		mousePos.y > currentSprite.getPosition().y && mousePos.y < (currentSprite.getPosition().y + height))
		setState(!currentState);
}

void UIButton::setState(bool which)
{
	currentState = which;
	if (currentState)
	{
		currentSprite = clicked;
		return;
	}
	currentSprite = normal;
}

File: C#/exercise 04/sfml.net/src/UIButton.cs
using SFML.Graphics;
using SFML.System;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace sfml.net.src
{
    public delegate void OnClickEvent();
    class UIButton
    {
        #region Fields

        private Sprite spriteNormal;
        private Sprite spriteClicked;
        private Sprite spriteCurrent;
        private bool currentState;
        private Texture textureNormal;
        private Texture textureClicked;
        public Sprite Sprite { get { return spriteCurrent; } }
        public bool State { get { return currentState; } }
        public OnClickEvent OnClickEvent;
        public bool IsInteractable { get; set; }

        #endregion


        #region Public

        public UIButton(string normalTexturePath, string clickedTexturePath, Vector2f uiPosition)
        {
            if ((textureNormal = new Texture(normalTexturePath)) == null)
                throw new Exception("Error loading texture");

            if ((textureClicked = new Texture(clickedTexturePath)) == null)
                throw new Exception("Error loading texture");

            OnClickEvent += () => { };
            spriteNormal = new Sprite(textureNormal);
            spriteClicked = new Sprite(textureClicked);
            spriteNormal.Position = uiPosition;
            spriteClicked.Position = uiPosition;
            IsInteractable = true;
            SetState(false);
        }

        public void CheckClick(Vector2f mousePos)
        {
            var width = spriteCurrent.GetGlobalBounds().Width;
            var height = spriteCurrent.GetGlobalBounds().Height;

            if (mousePos.X > spriteCurrent.Position.X && mousePos.X < (spriteCurrent.Position.X + width) &&
                mousePos.Y > spriteCurrent.Position.Y && mousePos.Y < (spriteCurrent.Position.Y + height))
                SetState(!currentState);
        }

        #endregion


        #region Private

        private void SetState(bool v)
        {
            if (IsInteractable)
            {
                currentState = v;
                if (currentState)
                {
                    spriteCurrent = spriteClicked;
                    OnClickEvent();
                }
                else
                    spriteCurrent = spriteNormal;
            }
        }

        #endregion
    }
}
